File: android/src/main/java/com/klippa/capacitorscanner/KlippaScannerSDKPlugin.kt
package com.klippa.capacitorscanner

import androidx.activity.result.ActivityResult
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.ActivityCallback
import com.getcapacitor.annotation.CapacitorPlugin
import com.klippa.scanner.ScannerFinishedReason
import com.klippa.scanner.ScannerSession
import com.klippa.scanner.ScannerSession.Companion.KLIPPA_ERROR
import com.klippa.scanner.ScannerSession.Companion.KLIPPA_RESULT
import com.klippa.scanner.model.Instructions
import com.klippa.scanner.model.KlippaCameraModes
import com.klippa.scanner.model.KlippaDPI
import com.klippa.scanner.model.KlippaDocumentMode
import com.klippa.scanner.model.KlippaError
import com.klippa.scanner.model.KlippaImageColor
import com.klippa.scanner.model.KlippaMultipleDocumentMode
import com.klippa.scanner.model.KlippaObjectDetectionModel
import com.klippa.scanner.model.KlippaOutputFormat
import com.klippa.scanner.model.KlippaPageFormat
import com.klippa.scanner.model.KlippaScannerResult
import com.klippa.scanner.model.KlippaSegmentedDocumentMode
import com.klippa.scanner.model.KlippaSingleDocumentMode
import com.klippa.scanner.model.KlippaSize
import com.klippa.scanner.storage.KlippaStorage
import org.json.JSONArray
import org.json.JSONObject

@CapacitorPlugin(name = "KlippaScannerSDKPlugin")
class KlippaScannerSDKPlugin : Plugin() {

    companion object {
        private const val E_ACTIVITY_DOES_NOT_EXIST = "E_ACTIVITY_DOES_NOT_EXIST"
        private const val E_MISSING_LICENSE = "E_MISSING_LICENSE"
        private const val E_FAILED_TO_SHOW_SESSION = "E_FAILED_TO_SHOW_SESSION"
        private const val E_CANCELED = "E_CANCELED"
    }

    private var _call: PluginCall? = null

    data object UnknownError: KlippaError {
        private fun readResolve(): Any = UnknownError
        override fun message(): String = "Unknown Error"
    }

    @PluginMethod
    fun purge(call: PluginCall) {
        activity?.let { 
            KlippaStorage.purge(it)
            call.resolve()
        } ?: run {
            call.reject("Activity doesn't exist", E_ACTIVITY_DOES_NOT_EXIST)
        }
    }

    @PluginMethod
    fun getCameraResult(call: PluginCall) {
        val config: JSONObject = call.data

        val license = config.getStringOrNull("license") ?: run {
            call.reject("Missing license", E_MISSING_LICENSE)
            return
        }

        activity?.let { activity ->
            try {
                val scannerSession = ScannerSession(license)

                // Global options
                config.getBooleanOrNull("defaultCrop")?.let {
                    scannerSession.menu.isCropEnabled = it
                }

                config.getIntOrNull("imageMaxWidth")?.let {
                    scannerSession.imageAttributes.resolutionMaxWidth = it
                }

                config.getIntOrNull("imageMaxHeight")?.let {
                    scannerSession.imageAttributes.resolutionMaxHeight = it
                }

                config.getIntOrNull("imageMaxQuality")?.let {
                    scannerSession.imageAttributes.outputQuality = it
                }

                config.getDoubleOrNull("previewDuration")?.let {
                    scannerSession.durations.previewDuration = it
                }

                // Model setup
                config.getJsonObjectOrNull("model")?.let { model ->
                    val fileName = model.getStringOrNull("fileName")
                    val modelLabels = model.getStringOrNull("modelLabels")

                    if (!fileName.isNullOrBlank() && !modelLabels.isNullOrBlank()) {
                        val objectDetectionModel = KlippaObjectDetectionModel().apply {
                            modelName = fileName
                            this.modelLabels = modelLabels
                        }
                        scannerSession.objectDetectionModel = objectDetectionModel
                    }
                }

                // Timer setup
                config.getJsonObjectOrNull("timer")?.let { timer ->
                    timer.getBooleanOrNull("allowed")?.let {
                        scannerSession.menu.allowTimer = it
                    }
                    
                    timer.getBooleanOrNull("enabled")?.let {
                        scannerSession.menu.isTimerEnabled = it
                    }
                    
                    timer.getDoubleOrNull("duration")?.let {
                        scannerSession.durations.timerDuration = it
                    }
                }

                // Crop padding setup
                config.getJsonObjectOrNull("cropPadding")?.let { cropPadding ->
                    val width = cropPadding.getIntOrNull("width") ?: 0
                    val height = cropPadding.getIntOrNull("height") ?: 0
                    scannerSession.imageAttributes.cropPadding = KlippaSize(width, height)
                }

                // Success setup
                config.getJsonObjectOrNull("success")?.let { success ->
                    success.getDoubleOrNull("previewDuration")?.let {
                        scannerSession.durations.successPreviewDuration = it
                    }
                }

                // Shutter button setup
                config.getJsonObjectOrNull("shutterButton")?.let { shutterButton ->
                    shutterButton.getBooleanOrNull("allowShutterButton")?.let {
                        scannerSession.shutterButton.allowShutterButton = it
                    }
                    
                    shutterButton.getBooleanOrNull("hideShutterButton")?.let {
                        scannerSession.shutterButton.hideShutterButton = it
                    }
                }

                // Storage and output
                config.getStringOrNull("storagePath")?.let {
                    scannerSession.imageAttributes.outputDirectory = it
                }

                config.getStringOrNull("outputFilename")?.let {
                    scannerSession.imageAttributes.outputFileName = it
                }

                // Default color
                config.getStringOrNull("defaultColor")?.let {
                    when (it) {
                        "original" -> scannerSession.imageAttributes.imageColorMode = KlippaImageColor.ORIGINAL
                        "enhanced" -> scannerSession.imageAttributes.imageColorMode = KlippaImageColor.ENHANCED
                        "grayscale" -> scannerSession.imageAttributes.imageColorMode = KlippaImageColor.GRAYSCALE
                        "blackAndWhite" -> scannerSession.imageAttributes.imageColorMode = KlippaImageColor.BLACK_AND_WHITE
                    }
                }

                // Output format
                config.getStringOrNull("outputFormat")?.let {
                    when (it) {
                        "jpeg" -> scannerSession.imageAttributes.outputFormat = KlippaOutputFormat.JPEG
                        "pdfSingle" -> scannerSession.imageAttributes.outputFormat = KlippaOutputFormat.PDF_SINGLE
                        "pdfMerged" -> scannerSession.imageAttributes.outputFormat = KlippaOutputFormat.PDF_MERGED
                        "png" -> scannerSession.imageAttributes.outputFormat = KlippaOutputFormat.PNG
                    }
                }

                config.getIntOrNull("imageLimit")?.let {
                    scannerSession.imageAttributes.imageLimit = it
                }

                config.getIntOrNull("imageMovingSensitivityAndroid")?.let {
                    scannerSession.imageAttributes.imageMovingSensitivity = it
                }

                config.getBooleanOrNull("performOnDeviceOCR")?.let {
                    scannerSession.imageAttributes.performOnDeviceOCR = it
                }

                config.getBooleanOrNull("storeImagesToCameraRoll")?.let {
                    scannerSession.imageAttributes.storeImagesToGallery = it
                }

                // Page format
                config.getStringOrNull("pageFormat")?.let {
                    when (it) {
                        "off" -> scannerSession.imageAttributes.pageFormat = KlippaPageFormat.OFF
                        "a3" -> scannerSession.imageAttributes.pageFormat = KlippaPageFormat.A3
                        "a4" -> scannerSession.imageAttributes.pageFormat = KlippaPageFormat.A4
                        "a5" -> scannerSession.imageAttributes.pageFormat = KlippaPageFormat.A5
                        "a6" -> scannerSession.imageAttributes.pageFormat = KlippaPageFormat.A6
                        "b4" -> scannerSession.imageAttributes.pageFormat = KlippaPageFormat.B4
                        "b5" -> scannerSession.imageAttributes.pageFormat = KlippaPageFormat.B5
                        "letter" -> scannerSession.imageAttributes.pageFormat = KlippaPageFormat.LETTER
                    }
                }

                // DPI setting
                config.getStringOrNull("dpi")?.let {
                    when (it) {
                        "auto" -> scannerSession.imageAttributes.dpi = KlippaDPI.AUTO
                        "dpi200" -> scannerSession.imageAttributes.dpi = KlippaDPI.DPI_200
                        "dpi300" -> scannerSession.imageAttributes.dpi = KlippaDPI.DPI_300
                    }
                }

                // Brightness thresholds
                config.getDoubleOrNull("brightnessLowerThreshold")?.let {
                    scannerSession.imageAttributes.brightnessLowerThreshold = it
                }

                config.getDoubleOrNull("brightnessUpperThreshold")?.let {
                    scannerSession.imageAttributes.brightnessUpperThreshold = it
                }

                // Menu options
                config.getBooleanOrNull("shouldGoToReviewScreenWhenImageLimitReached")?.let {
                    scannerSession.menu.shouldGoToReviewScreenWhenImageLimitReached = it
                }

                config.getBooleanOrNull("userShouldAcceptResultToContinue")?.let {
                    scannerSession.menu.userShouldAcceptResultToContinue = it
                }

                config.getBooleanOrNull("userCanRotateImage")?.let {
                    scannerSession.menu.userCanRotateImage = it
                }

                config.getBooleanOrNull("userCanCropManually")?.let {
                    scannerSession.menu.userCanCropManually = it
                }

                config.getBooleanOrNull("userCanChangeColorSetting")?.let {
                    scannerSession.menu.userCanChangeColorSetting = it
                }

                config.getBooleanOrNull("userCanPickMediaFromStorage")?.let {
                    scannerSession.menu.userCanPickMediaFromStorage = it
                }

                config.getBooleanOrNull("shouldGoToReviewScreenOnFinishPressed")?.let {
                    scannerSession.menu.shouldGoToReviewScreenOnFinishPressed = it
                }

                // Camera modes setup
                setupCameraModes(config, scannerSession)

                val intent = scannerSession.getIntent(activity)
                _call = call
                startActivityForResult(call, intent, "klippaScannerResult")

            } catch (e: Exception) {
                call.reject(
                    "Could not launch scanner session: ${e.message}",
                    E_FAILED_TO_SHOW_SESSION
                )
            }
        } ?: run {
            call.reject("Activity doesn't exist", E_ACTIVITY_DOES_NOT_EXIST)
        }
    }

    private fun setupCameraModes(config: JSONObject, scannerSession: ScannerSession) {
        val modes: MutableList<KlippaDocumentMode> = mutableListOf()

        config.getJsonObjectOrNull("cameraModeSingle")?.let { cameraModeSingle ->
            val singleDocumentMode = KlippaSingleDocumentMode()
            cameraModeSingle.getStringOrNull("name")?.let { name ->
                singleDocumentMode.name = name
            }
            cameraModeSingle.getStringOrNull("message")?.let { message ->
                singleDocumentMode.instructions = Instructions(singleDocumentMode.name, message)
            }
            modes.add(singleDocumentMode)
        }

        config.getJsonObjectOrNull("cameraModeMulti")?.let { cameraModeMulti ->
            val multiDocumentMode = KlippaMultipleDocumentMode()
            cameraModeMulti.getStringOrNull("name")?.let { name ->
                multiDocumentMode.name = name
            }
            cameraModeMulti.getStringOrNull("message")?.let { message ->
                multiDocumentMode.instructions = Instructions(multiDocumentMode.name, message)
            }
            modes.add(multiDocumentMode)
        }

        config.getJsonObjectOrNull("cameraModeSegmented")?.let { cameraModeSegmented ->
            val segmentedDocumentMode = KlippaSegmentedDocumentMode()
            cameraModeSegmented.getStringOrNull("name")?.let { name ->
                segmentedDocumentMode.name = name
            }
            cameraModeSegmented.getStringOrNull("message")?.let { message ->
                segmentedDocumentMode.instructions = Instructions(segmentedDocumentMode.name, message)
            }
            modes.add(segmentedDocumentMode)
        }

        if (modes.isNotEmpty()) {
            var index = 0
            config.getIntOrNull("startingIndex")?.let {
                index = it
            }
            val cameraModes = KlippaCameraModes(
                modes = modes,
                startingIndex = index
            )
            scannerSession.cameraModes = cameraModes
        }
    }

    @Suppress("UNUSED")
    @ActivityCallback
    private fun klippaScannerResult(call: PluginCall, result: ActivityResult) {
        val reason = ScannerFinishedReason.mapResultCode(result.resultCode)

        val data = result.data

        when (reason) {
            ScannerFinishedReason.FINISHED -> {
                val result = data?.serializable<KlippaScannerResult>(KLIPPA_RESULT) ?: run {
                    klippaScannerDidFailWithError(UnknownError)
                    return
                }
                klippaScannerDidFinishScanningWithResult(result)
            }
            ScannerFinishedReason.ERROR -> {
                val error = data?.serializable<KlippaError>(KLIPPA_ERROR) ?: run {
                    klippaScannerDidFailWithError(UnknownError)
                    return
                }
                klippaScannerDidFailWithError(error)
            }
            ScannerFinishedReason.CANCELED -> {
                klippaScannerDidCancel()
            }
            else -> klippaScannerDidFailWithError(UnknownError)
        }

    }

    private fun klippaScannerDidFinishScanningWithResult(result: KlippaScannerResult) {
        val images: MutableList<Map<String, String>> = mutableListOf()

        for (image in result.results) {
            val imageDict = mapOf("filePath" to image.location)
            images.add(imageDict)
        }

        val cropEnabled: Boolean = result.cropEnabled
        val timerEnabled: Boolean = result.timerEnabled

        val singleDocumentModeInstructionsDismissed = result.dismissedInstructions["SINGLE_DOCUMENT"] ?: false
        val multiDocumentModeInstructionsDismissed = result.dismissedInstructions["MULTIPLE_DOCUMENT"] ?: false
        val segmentedDocumentModeInstructionsDismissed = result.dismissedInstructions["SEGMENTED_DOCUMENT"] ?: false

        val imagesAsJSON = JSONArray(images)

        val ret = JSObject()
        ret.put("images", imagesAsJSON)
        ret.put("crop", cropEnabled)
        ret.put("timerEnabled", timerEnabled)
        ret.put("singleDocumentModeInstructionsDismissed", singleDocumentModeInstructionsDismissed)
        ret.put("multiDocumentModeInstructionsDismissed", multiDocumentModeInstructionsDismissed)
        ret.put("segmentedDocumentModeInstructionsDismissed", segmentedDocumentModeInstructionsDismissed)

        _call?.resolve(ret)
        _call = null
    }

    private fun klippaScannerDidFailWithError(error: KlippaError) {
        _call?.reject("Scanner canceled with error: ${error.message()}", E_CANCELED)
        _call = null
    }

    private fun klippaScannerDidCancel() {
        _call?.reject("Scanner was canceled", E_CANCELED)
        _call = null
    }

}

File: ios/Plugin/KlippaScannerSDKPlugin.swift
import Foundation
import Capacitor
import KlippaScanner

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(KlippaScannerSDKPlugin)
public class KlippaScannerSDKPlugin: CAPPlugin {

    private let E_ERROR = "E_ERROR"
    private let E_CANCELED = "E_CANCELED"
    private let E_MISSING_LICENSE = "E_MISSING_LICENSE"
    private let E_FAILED_TO_SHOW_SESSION = "E_FAILED_TO_SHOW_SESSION"
    private let E_UNKNOWN_ERROR = "E_UNKNOWN_ERROR"

    private var _call: CAPPluginCall? = nil

    // MARK: Setup Timer
    private func setupTimer(_ config: [AnyHashable : Any], _ builder: KlippaScannerBuilder) {
        guard let timer = config["timer"] as? [String: Any] else {
            return
        }

        if let isTimerAllowed = timer["allowed"] as? Bool {
            builder.klippaMenu.allowTimer = isTimerAllowed
        }

        if let isTimerEnabled = timer["enabled"] as? Bool {
            builder.klippaMenu.isTimerEnabled = isTimerEnabled
        }

        if let timerDuration = timer["duration"] as? Double {
            builder.klippaDurations.timerDuration = timerDuration
        }
    }

    // MARK: Purge
    @objc func purge(_ call: CAPPluginCall) {
        KlippaScannerStorage.purge()
        call.resolve()
    }

    //MARK: Setup Colors
    private func setupColors(_ config: [AnyHashable : Any], _ builder: KlippaScannerBuilder) {
        if let primaryColor = config["primaryColor"] as? String {
            builder.klippaColors.primaryColor = hexColorToUIColor(hex: primaryColor)
        }

        if let accentColor = config["accentColor"] as? String {
            builder.klippaColors.accentColor = hexColorToUIColor(hex: accentColor)
        }

        if let secondaryColor = config["secondaryColor"] as? String {
            builder.klippaColors.secondaryColor = hexColorToUIColor(hex: secondaryColor)
        }

        if let overlayColorAlpha = config["overlayColorAlpha"] as? CGFloat {
            builder.klippaColors.overlayColorAlpha = overlayColorAlpha
        }

        if let warningBackgroundColor = config["warningBackgroundColor"] as? String {
            builder.klippaColors.warningBackgroundColor = hexColorToUIColor(hex: warningBackgroundColor)
        }

        if let warningTextColor = config["warningTextColor"] as? String {
            builder.klippaColors.warningTextColor = hexColorToUIColor(hex: warningTextColor)
        }

        if let iconEnabledColor = config["iconEnabledColor"] as? String {
            builder.klippaColors.iconEnabledColor = hexColorToUIColor(hex: iconEnabledColor)
        }

        if let iconDisabledColor = config["iconDisabledColor"] as? String {
            builder.klippaColors.iconDisabledColor = hexColorToUIColor(hex: iconDisabledColor)
        }

        if let buttonWithIconForegroundColor = config["buttonWithIconForegroundColor"] as? String {
            builder.klippaColors.buttonWithIconForegroundColor = hexColorToUIColor(hex: buttonWithIconForegroundColor)
        }

        if let buttonWithIconBackgroundColor = config["buttonWithIconBackgroundColor"] as? String {
            builder.klippaColors.buttonWithIconBackgroundColor = hexColorToUIColor(hex: buttonWithIconBackgroundColor)
        }

        if let primaryActionForegroundColor = config["primaryActionForegroundColor"] as? String {
            builder.klippaColors.primaryActionForegroundColor = hexColorToUIColor(hex: primaryActionForegroundColor)
        }

        if let primaryActionBackgroundColor = config["primaryActionBackgroundColor"] as? String {
            builder.klippaColors.primaryActionBackgroundColor = hexColorToUIColor(hex: primaryActionBackgroundColor)
        }

        if let defaultColor = config["defaultColor"] as? String {
            let color = switch defaultColor {
            case "grayscale":
                KlippaImageColor.grayscale
            case "enhanced":
                KlippaImageColor.enhanced
            case "blackAndWhite":
                KlippaImageColor.blackAndWhite
            default:
                KlippaImageColor.original
            }
            builder.klippaColors.imageColor = color
        }
    }

    // MARK: Setup Model
    private func setupModel(_ config: [AnyHashable : Any], _ builder: KlippaScannerBuilder) {
        if let model = config["model"] as? [String: String] {
            if let modelFile = model["fileName"], let modelLabels = model["modelLabels"] {
                builder.klippaObjectDetectionModel = KlippaObjectDetectionModel(
                    modelFile: modelFile,
                    modelLabels: modelLabels
                )
            }
        }
    }

    // MARK: Setup Camera Modes
    private func setupCameraModes(_ config: [AnyHashable : Any], _ builder: KlippaScannerBuilder) {
        var modes = [KlippaDocumentMode]()

        if let singleCameraMode = config["cameraModeSingle"] as? [String: String] {
            let single = KlippaSingleDocumentMode()

            if let name = singleCameraMode["name"] {
                single.name = name
            }

            if let message = singleCameraMode["message"] ?? single.instructions?.message {
                let image = singleCameraMode["image"]
                single.instructions = Instructions(
                    title: single.name,
                    message: message,
                    image: image ?? KlippaSingleDocumentMode.image
                )
            }
            modes.append(single)
        }

        if let multiCameraMode = config["cameraModeMulti"] as? [String: String] {
            let multi = KlippaMultipleDocumentMode()

            if let name = multiCameraMode["name"] {
                multi.name = name
            }

            if let message = multiCameraMode["message"] ?? multi.instructions?.message {
                let image = multiCameraMode["image"]
                multi.instructions = Instructions(
                    title: multi.name,
                    message: message,
                    image: image ?? KlippaMultipleDocumentMode.image
                )
            }
            modes.append(multi)
        }

        if let segmentedCameraMode = config["cameraModeSegmented"] as? [String: String] {
            let segmented = KlippaSegmentedDocumentMode()

            if let name = segmentedCameraMode["name"] {
                segmented.name = name
            }

            if let message = segmentedCameraMode["message"] ?? segmented.instructions?.message {
                let image = segmentedCameraMode["image"]
                segmented.instructions = Instructions(
                    title: segmented.name,
                    message: message,
                    image: image ?? KlippaSegmentedDocumentMode.image
                )
            }
            modes.append(segmented)
        }

        if modes.isEmpty {
            return
        }

        let index = config["startingIndex"] as? Int ?? 0

        builder.klippaCameraModes = KlippaCameraModes(
            modes: modes,
            startingIndex: index
        )
    }

    //MARK: Setup Messages
    private func setupMessages(_ config: [AnyHashable : Any], _ builder: KlippaScannerBuilder) {
        if let imageTooBrightMessage = config["imageTooBrightMessage"] as? String {
            builder.klippaMessages.imageTooBrightMessage = imageTooBrightMessage
        }

        if let imageTooDarkMessage = config["imageTooDarkMessage"] as? String {
            builder.klippaMessages.imageTooDarkMessage = imageTooDarkMessage
        }

        if let imageColorOriginalText = config["imageColorOriginalText"] as? String {
            builder.klippaButtonTexts.imageColorOriginalText = imageColorOriginalText
        }

        if let imageColorGrayscaleText = config["imageColorGrayScaleText"] as? String {
            builder.klippaButtonTexts.imageColorGrayscaleText = imageColorGrayscaleText
        }

        if let imageColorEnhancedText = config["imageColorEnhancedText"] as? String {
            builder.klippaButtonTexts.imageColorEnhancedText = imageColorEnhancedText
        }

        if let imageColorBlackAndWhiteText = config["imageColorBlackAndWhiteText"] as? String {
            builder.klippaButtonTexts.imageColorBlackAndWhiteText = imageColorBlackAndWhiteText
        }

        if let continueButtonText = config["continueButtonText"] as? String {
            builder.klippaButtonTexts.continueButtonText = continueButtonText
        }

        if let imageLimitReachedMessage = config["imageLimitReachedMessage"] as? String {
            builder.klippaMessages.imageLimitReachedMessage = imageLimitReachedMessage
        }

        if let deleteButtonText = config["deleteButtonText"] as? String {
            builder.klippaButtonTexts.deleteButtonText = deleteButtonText
        }

        if let retakeButtonText = config["retakeButtonText"] as? String {
            builder.klippaButtonTexts.retakeButtonText = retakeButtonText
        }

        if let cancelButtonText = config["cancelButtonText"] as? String {
            builder.klippaButtonTexts.cancelButtonText = cancelButtonText
        }

        if let cancelAndDeleteImagesButtonText = config["cancelAndDeleteImagesButtonText"] as? String {
            builder.klippaButtonTexts.cancelAndDeleteImagesButtonText = cancelAndDeleteImagesButtonText
        }

        if let cropEditButtonText = config["cropEditButtonText"] as? String {
            builder.klippaButtonTexts.cropEditButtonText = cropEditButtonText
        }

        if let filterEditButtonText = config["filterEditButtonText"] as? String {
            builder.klippaButtonTexts.filterEditButtonText = filterEditButtonText
        }

        if let rotateEditButtonText = config["rotateEditButtonText"] as? String {
            builder.klippaButtonTexts.rotateEditButtonText = rotateEditButtonText
        }

        if let deleteEditButtonText = config["deleteEditButtonText"] as? String {
            builder.klippaButtonTexts.deleteEditButtonText = deleteEditButtonText
        }

        if let cancelCropButtonText = config["cancelCropButtonText"] as? String {
            builder.klippaButtonTexts.cancelCropButtonText = cancelCropButtonText
        }

        if let expandCropButtonText = config["expandCropButtonText"] as? String {
            builder.klippaButtonTexts.expandCropButtonText = expandCropButtonText
        }

        if let saveCropButtonText = config["saveCropButtonText"] as? String {
            builder.klippaButtonTexts.saveCropButtonText = saveCropButtonText
        }

        if let autoCaptureButtonText = config["autoCaptureButtonText"] as? String {
            builder.klippaButtonTexts.autoCaptureButtonText = autoCaptureButtonText
        }

        if let manualButtonText = config["manualButtonText"] as? String {
            builder.klippaButtonTexts.manualButtonText = manualButtonText
        }

        if let deleteOptionsButtonText = config["deleteOptionsButtonText"] as? String {
            builder.klippaButtonTexts.deleteOptionsButtonText = deleteOptionsButtonText
        }

        if let cancelConfirmationMessage = config["cancelConfirmationMessage"] as? String {
            builder.klippaMessages.cancelConfirmationMessage = cancelConfirmationMessage
        }

        if let segmentedModeImageCountMessage = config["segmentedModeImageCountMessage"] as? String {
            builder.klippaMessages.segmentedModeImageCountMessage = segmentedModeImageCountMessage
        }

        if let moveCloserMessage = config["moveCloserMessage"] as? String {
            builder.klippaMessages.moveCloserMessage = moveCloserMessage
        }

        if let imageMovingMessage = config["imageMovingMessage"] as? String {
            builder.klippaMessages.imageMovingMessage = imageMovingMessage
        }
    }

    // MARK: Setup Shutter Button
    private func setupShutterButton(_ config: [AnyHashable : Any], _ builder: KlippaScannerBuilder) {
        if let shutterButton = config["shutterButton"] as? [String: Bool] {
            if let allowShutterButton = shutterButton["allowShutterButton"] {
                builder.klippaShutterButton.allowShutterButton = allowShutterButton
            }

            if let hideShutterButton = shutterButton["hideShutterButton"] {
                builder.klippaShutterButton.hideShutterButton = hideShutterButton
            }
        }
    }

    // MARK: Setup Image Settings
    private func setupImageSettings(_ config: [AnyHashable : Any], _ builder: KlippaScannerBuilder) {
        if let imageMaxWidth = config["imageMaxWidth"] as? CGFloat {
            builder.klippaImageAttributes.imageMaxWidth = imageMaxWidth
        }

        if let imageMaxHeight = config["imageMaxHeight"] as? CGFloat {
            builder.klippaImageAttributes.imageMaxHeight = imageMaxHeight
        }

        if let imageMaxQuality = config["imageMaxQuality"] as? CGFloat {
            builder.klippaImageAttributes.imageMaxQuality = imageMaxQuality
        }

        if let cropPadding = config["cropPadding"] as? [String: Double] {
            if let width = cropPadding["width"] {
                builder.klippaImageAttributes.cropPadding.width = width
            }
            if let height = cropPadding["height"] {
                builder.klippaImageAttributes.cropPadding.height = height
            }
        }

        if let imageLimit = config["imageLimit"] as? Int {
            builder.klippaImageAttributes.imageLimit = imageLimit
        }

        if let storeImagesToCameraRoll = config["storeImagesToCameraRoll"] as? Bool {
            builder.klippaImageAttributes.storeImagesToCameraRoll = storeImagesToCameraRoll
        }

        if let imageMovingSensitivity = config["imageMovingSensitivityIOS"] as? CGFloat {
            builder.klippaImageAttributes.imageMovingSensitivity = imageMovingSensitivity
        }

        if let outputFormat = config["outputFormat"] as? String {
            switch outputFormat {
            case "jpeg":
                builder.klippaImageAttributes.outputFormat = .jpeg
            case "pdfSingle":
                builder.klippaImageAttributes.outputFormat = .pdfSingle
            case "pdfMerged":
                builder.klippaImageAttributes.outputFormat = .pdfMerged
            case "png":
                builder.klippaImageAttributes.outputFormat = .png
            default:
                builder.klippaImageAttributes.outputFormat = .jpeg
            }
        }

        if let performOnDeviceOCR = config["performOnDeviceOCR"] as? Bool {
            builder.klippaImageAttributes.performOnDeviceOCR = performOnDeviceOCR
        }

        if let brightnessLowerThreshold = config["brightnessLowerThreshold"] as? Double {
            builder.klippaImageAttributes.brightnessLowerThreshold = brightnessLowerThreshold
        }

        if let brightnessUpperThreshold = config["brightnessUpperThreshold"] as? Double {
            builder.klippaImageAttributes.brightnessUpperThreshold = brightnessUpperThreshold
        }

        if let pageFormat = config["pageFormat"] as? String {
            switch pageFormat {
            case "off":
                builder.klippaImageAttributes.pageFormat = .off
            case "a3":
                builder.klippaImageAttributes.pageFormat = .a3
            case "a4":
                builder.klippaImageAttributes.pageFormat = .a4
            case "a5":
                builder.klippaImageAttributes.pageFormat = .a5
            case "a6":
                builder.klippaImageAttributes.pageFormat = .a6
            case "b4":
                builder.klippaImageAttributes.pageFormat = .b4
            case "b5":
                builder.klippaImageAttributes.pageFormat = .b5
            case "letter":
                builder.klippaImageAttributes.pageFormat = .letter
            default:
                builder.klippaImageAttributes.pageFormat = .off
            }
        }

        if let dpi = config["dpi"] as? String {
            switch dpi {
            case "auto":
                builder.klippaImageAttributes.dpi = .auto
            case "dpi200":
                builder.klippaImageAttributes.dpi = .dpi200
            case "dpi300":
                builder.klippaImageAttributes.dpi = .dpi300
            default:
                builder.klippaImageAttributes.dpi = .auto
            }
        }
    }

    // MARK: Setup Menu
    private func setupMenu(_ config: [AnyHashable : Any], _ builder: KlippaScannerBuilder) {
        if let isCropEnabled = config["defaultCrop"] as? Bool {
            builder.klippaMenu.isCropEnabled = isCropEnabled
        }

        if let shouldGoToReviewScreenWhenImageLimitReached = config["shouldGoToReviewScreenWhenImageLimitReached"] as? Bool {
            builder.klippaMenu.shouldGoToReviewScreenWhenImageLimitReached = shouldGoToReviewScreenWhenImageLimitReached
        }

        if let userShouldAcceptResultToContinue = config["userShouldAcceptResultToContinue"] as? Bool {
            builder.klippaMenu.userShouldAcceptResultToContinue = userShouldAcceptResultToContinue
        }

        if let userCanRotateImage = config["userCanRotateImage"] as? Bool {
            builder.klippaMenu.userCanRotateImage = userCanRotateImage
        }

        if let userCanCropManually = config["userCanCropManually"] as? Bool {
            builder.klippaMenu.userCanCropManually = userCanCropManually
        }

        if let userCanChangeColorSetting = config["userCanChangeColorSetting"] as? Bool {
            builder.klippaMenu.userCanChangeColorSetting = userCanChangeColorSetting
        }

        if let isViewFinderEnabled = config["isViewFinderEnabled"] as? Bool {
            builder.klippaMenu.isViewFinderEnabled = isViewFinderEnabled
        }

        if let userCanPickMediaFromStorage = config["userCanPickMediaFromStorage"] as? Bool {
            builder.klippaMenu.userCanPickMediaFromStorage = userCanPickMediaFromStorage
        }

        if let shouldGoToReviewScreenOnFinishPressed = config["shouldGoToReviewScreenOnFinishPressed"] as? Bool {
            builder.klippaMenu.shouldGoToReviewScreenOnFinishPressed = shouldGoToReviewScreenOnFinishPressed
        }
    }

    // MARK: Setup Success
    private func setupSuccess(_ config: [AnyHashable : Any], _ builder: KlippaScannerBuilder) {
        if let previewDuration = config["previewDuration"] as? Double {
            builder.klippaDurations.previewDuration = previewDuration
        }

        guard let success = config["success"] as? [String: Any] else {
            return
        }

        if let successPreviewDuration = success["previewDuration"] as? Double {
            builder.klippaDurations.successPreviewDuration = successPreviewDuration
        }

        if let successMessage = success["message"] as? String {
            builder.klippaMessages.successMessage = successMessage
        }

    }

    // MARK: Get Camera Result
    @objc func getCameraResult(_ call: CAPPluginCall) {
        guard let config = call.options else {
            call.reject("CameraConfig was not supplied correctly.", E_UNKNOWN_ERROR)
            return
        }

        guard let license = config["license"] as? String else {
            call.reject("Missing license", E_MISSING_LICENSE)
            return
        }

        let builder = KlippaScannerBuilder(
            builderDelegate: self,
            license: license
        )

        setupMenu(config, builder)

        setupImageSettings(config, builder)

        setupTimer(config, builder)

        setupSuccess(config, builder)

        setupShutterButton(config, builder)

        setupMessages(config, builder)

        setupColors(config, builder)

        setupModel(config, builder)

        setupCameraModes(config, builder)

        _call = call

        DispatchQueue.main.async {
            let result = builder.build()
            let rootViewController = UIApplication.shared.windows.last!.rootViewController!

            switch result {
            case .success(let controller):
                controller.modalPresentationStyle = .fullScreen
                rootViewController.present(controller, animated: true, completion: nil)
            case .failure(let error):
                call.reject("error: \(error)", self.E_FAILED_TO_SHOW_SESSION, error)
            }
        }
    }

    // MARK: Hex to UIColor
    private func hexColorToUIColor(hex: String) -> UIColor {
        let r, g, b, a: CGFloat

        let start = hex.index(hex.startIndex, offsetBy: 1)
        let hexColor = String(hex[start...])

        if hexColor.count == 8 {
            let scanner = Scanner(string: hexColor)
            var hexNumber: UInt64 = 0

            if scanner.scanHexInt64(&hexNumber) {
                a = CGFloat((hexNumber & 0xff000000) >> 24) / 255
                r = CGFloat((hexNumber & 0x00ff0000) >> 16) / 255
                g = CGFloat((hexNumber & 0x0000ff00) >> 8) / 255
                b = CGFloat(hexNumber & 0x000000ff) / 255

                return UIColor(red: r, green: g, blue: b, alpha: a)
            }
        }

        // Fallback to 6-character hex format
        var cString: String = hex.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()

        if (cString.hasPrefix("#")) {
            cString.remove(at: cString.startIndex)
        }

        if ((cString.count) != 6) {
            return UIColor.gray
        }

        var rgbValue: UInt64 = 0
        Scanner(string: cString).scanHexInt64(&rgbValue)

        return UIColor(
            red: CGFloat((rgbValue & 0xFF0000) >> 16) / 255.0,
            green: CGFloat((rgbValue & 0x00FF00) >> 8) / 255.0,
            blue: CGFloat(rgbValue & 0x0000FF) / 255.0,
            alpha: CGFloat(1.0)
        )
    }
}

// MARK: Scanner Delegate
extension KlippaScannerSDKPlugin: KlippaScannerDelegate {

    public func klippaScannerDidFinishScanningWithResult(result: KlippaScanner.KlippaScannerResult) {
        var images: [[String: String]] = []
        for image in result.results {
            let imageDict = ["filePath" : image.path]
            images.append(imageDict)
        }

        let singleDocumentModeInstructionsDismissed = result.dismissedInstructions[DocModeType.singleDocument.name] ?? false
        let multiDocumentModeInstructionsDismissed = result.dismissedInstructions[DocModeType.multipleDocument.name] ?? false
        let segmentedDocumentModeInstructionsDismissed = result.dismissedInstructions[DocModeType.segmentedDocument.name] ?? false

        let resultDict = [
            "images" : images,
            "crop": result.cropEnabled,
            "timerEnabled" : result.timerEnabled,
            "singleDocumentModeInstructionsDismissed": singleDocumentModeInstructionsDismissed,
            "multiDocumentModeInstructionsDismissed": multiDocumentModeInstructionsDismissed,
            "segmentedDocumentModeInstructionsDismissed": segmentedDocumentModeInstructionsDismissed
        ] as [String : Any]

        _call?.resolve(resultDict)
        _call = nil
    }

    public func klippaScannerDidCancel() {
        _call?.reject(E_CANCELED)
        _call = nil
    }

    public func klippaScannerDidFailWithError(error: Error) {
        _call?.reject(error.localizedDescription, E_ERROR, error)
        _call = nil
    }

}

